runner/execer/os/os_execer.go

```diff
@@ -239,20 +239,28 @@ PSLIST=$(ps -e -o pid= -o pgid= -o ppid= -o rss= | tr '\n' ';' | sed 's,;$,,')
 echo "
 all_processes=dict()
 related_processes=dict()
-id=None
+process_groups=dict()
+parent_processes=dict()
+children = []
+proc_group_id = None
 total=0
 for line in \"$PSLIST\".split(';'):
   pid, pgid, ppid, rss = tuple(line.split())
   all_processes[pid] = {'pgid': pgid, 'ppid': ppid, 'rss': rss}
-
-for pid, proc in all_processes.items():
-  if pid == \"$PID\" or proc['pgid'] == \"$PID\" or proc['ppid'] == \"$PID\":
-    related_processes[pid] = proc
-
-for rel_pid, _ in related_processes.items():
-  for pid, proc in all_processes.items():
-    if proc['pgid'] == rel_pid or proc['ppid'] == rel_pid:
-      related_processes[pid] = proc
+  process_groups.setdefault(pgid, []).append(pid)
+  parent_processes.setdefault(ppid, []).append(pid)
+  if pid == \"$PID\":
+    proc_group_id = pgid
+
+# Add all processes from pid's process group to related_processes
+for pid in process_groups.setdefault(proc_group_id, []):
+  related_processes[pid] = all_processes[pid]
```

Comment:
Still have to go through parent_processes here to catch the proc['ppid'] == pid case we were trying to add

Reply:
Do we? We should get that in line 263

Reply:
That's checking pids that were already in related procs, which we only added by pgid at 256-257. Won't the procs that set their own pgid (but still will have a chain to the PID through ppid) get left out?

Reply:
$PID is in related_processes too, so when pid == $PID, we should get any procs chained through ppid

Reply:
Ah, right, nevermind